wp/lib/bap_wp/src/output.ml

```diff
@@ -37,31 +37,42 @@ let format_mem_model (fmt : Format.formatter) (mem_model : mem_model) : unit =
   |> List.iter ~f:(fun (key, data) ->
       Format.fprintf fmt "\t\t%s |-> %s ;\n"
         (Expr.to_string key) (Constr.expr_to_hex data));
-  Format.fprintf fmt "\t\telse |-> %s]\n" (Constr.expr_to_hex mem_model.default)
+  if (Z3.BitVector.is_bv_numeral mem_model.default)
+  then Format.fprintf fmt "\t\telse |-> %s]\n" (Constr.expr_to_hex mem_model.default)
+  else Format.fprintf fmt "%s]\n" (Expr.to_string mem_model.default)
 
 (** [extract_array] takes a z3 expression that is a seqeunce of store and converts it into
     a mem_model, which consists of a key/value association list and a default value *)
 
 let extract_array (e : Constr.z3_expr) : mem_model =
   let rec extract_array' (partial_map : (Constr.z3_expr * Constr.z3_expr) list) (e : Constr.z3_expr) : mem_model =
     let numargs = Z3.Expr.get_num_args e in
-    let args = Z3.Expr.get_args e in
-    let f_decl = Z3.Expr.get_func_decl e in
-    let f_name = Z3.FuncDecl.get_name f_decl |> Z3.Symbol.to_string in
-    if Int.(numargs = 3) && String.(f_name = "store") then begin
-      let next_arr = List.nth_exn args 0 in
-      let key = List.nth_exn args 1 in
-      let value = List.nth_exn args 2 in
-      extract_array' ((key , value) :: partial_map) next_arr
-    end
-    else if Int.(numargs = 1) && String.(f_name = "const") then begin
-      let key = List.nth_exn args 0 in
-      { default = key; model = List.rev partial_map }
-    end
-    else begin
-      warning "Unexpected case destructing Z3 array: %s" (Z3.Expr.to_string e);
-      { default = e ; model = partial_map }
+    (* FIXME: Better condition for detecting lambda term. *)
+    if numargs <= 3 then begin
```

Comment:
I haven't tested it, but does `Z3Array.is_array` not work here?

Reply:
That is a very good idea. It might not, but I should try. I couldn't find an is_lambda function

Reply:
Yes this worked. Nice.

Reply:
Hmm. This goofs a little with the output when in Z3 mode
```
        mem_orig |-> [
                #x000105fc |-> 0x20 ;
                #x000105fd |-> 0x20 ;
                #x000105fe |-> 0x40 ;
                #x000105ff |-> 0x00 ;
                #x00402020 |-> 0x40 ;
                #x00402021 |-> 0x00 ;
                #x00402022 |-> 0x00 ;
                #x00402023 |-> 0x00 ;
                else |-> #x10]
```

Reply:
I'm a bit confused here. What changed?

Reply:
else |-> #x10] I think this used to be  `else |-> 0x10]` 
I suspect the constant case is not triggering is_array check so it's going through the other print statement

Reply:
We could use Expr.to_string on everything rather than Expr.expr_to_hex. Then I could remove the if statement also

Reply:
Okay, I see it now! Not sure what's best, so I'll leave it to your judgment here. I do have a slight preference for `0x10` over `#x10` though.